geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloIndexAdapter.scala

```diff
@@ -0,0 +1,452 @@
+          update.add(cf)
+          localityGroups.put(key, update)
+        } else {
+          localityGroups.put(key, Collections.singleton(cf))
+        }
+      }
+
+      groups.apply(index.sft).foreach { case (k, _) => addGroup(new Text(k)) }
+
+      if (localityGroups != existingGroups) {
+        tableOps.setLocalityGroups(table, localityGroups)
+      }
+
+      // enable block cache
+      tableOps.setProperty(table, Property.TABLE_BLOCKCACHE_ENABLED.getKey, "true")
+
+      if (index.name == IdIndex.name) {
+        // enable the row functor as the feature ID is stored in the Row ID
+        tableOps.setProperty(table, Property.TABLE_BLOOM_KEY_FUNCTOR.getKey, classOf[RowFunctor].getCanonicalName)
+        tableOps.setProperty(table, Property.TABLE_BLOOM_ENABLED.getKey, "true")
+      }
+    }
+  }
+
+  // noinspection ScalaDeprecation
+  override def deleteTables(tables: Seq[String]): Unit = {
+    tables.par.foreach { table =>
+      if (tableOps.exists(table)) {
+        if (ds.connector.isInstanceOf[org.apache.accumulo.core.client.mock.MockConnector]) {
+          // we need to synchronize deleting of tables in mock accumulo as it's not thread safe
+          ds.connector.synchronized(tableOps.delete(table))
+        } else {
+          tableOps.delete(table)
+        }
+      }
+    }
+  }
+
+  override def clearTables(tables: Seq[String], prefix: Option[Array[Byte]]): Unit = {
+    val auths = ds.auths // get the auths once up front
+    tables.par.foreach { table =>
+      if (tableOps.exists(table)) {
+        val config = GeoMesaBatchWriterConfig().setMaxWriteThreads(ds.config.writeThreads)
+        WithClose(ds.connector.createBatchDeleter(table, auths, ds.config.queryThreads, config)) { deleter =>
+          val range = prefix.map(p => Range.prefix(new Text(p))).getOrElse(new Range())
+          deleter.setRanges(Collections.singletonList(range))
+          deleter.delete()
+        }
+      }
+    }
+  }
+
+  override def createQueryPlan(strategy: QueryStrategy): AccumuloQueryPlan = {
+    import org.locationtech.geomesa.index.conf.QueryHints.RichHints
+
+    if (strategy.ranges.isEmpty) { EmptyPlan(strategy.filter) } else {
+      val QueryStrategy(filter, byteRanges, _, _, ecql, hints, _) = strategy
+      val index = filter.index
+      // index api defines empty start/end for open-ended range - in accumulo, it's indicated with null
+      // index api defines start row inclusive, end row exclusive
+      val ranges = byteRanges.map {
+        case BoundedByteRange(start, end) =>
+            val startKey = if (start.length == 0) { null } else { new Key(new Text(start)) }
+            val endKey = if (end.length == 0) { null } else { new Key(new Text(end)) }
+            new Range(startKey, true, endKey, false)
+
+        case SingleRowByteRange(row) =>
+          new Range(new Text(row))
+      }
+      val numThreads = if (index.name == IdIndex.name) { ds.config.recordThreads } else { ds.config.queryThreads }
```

Comment:
Interesting...  The number of query threads for the recordThreads can be different?!

Reply:
yeah, we've had this for a long time... it's used for id lookups and join queries - since each scan to the id index is a single row, the thread to performance ratio is different than against say the z3 index. I forget now if it's better to have it higher or lower, but I believe lower works better.